crates/oxc_semantic/src/builder.rs

```diff
@@ -520,18 +520,48 @@ impl<'a> SemanticBuilder<'a> {
     }
 
     fn reference_identifier(&mut self, ident: &IdentifierReference) {
-        let flag = if matches!(
-            self.nodes.parent_kind(self.current_node_id),
-            Some(AstKind::SimpleAssignmentTarget(_) | AstKind::AssignmentTarget(_))
-        ) {
-            ReferenceFlag::write()
-        } else {
-            ReferenceFlag::read()
-        };
+        let flag = self.resolve_reference_usages();
         let reference = Reference::new(ident.span, ident.name.clone(), flag);
         self.declare_reference(reference);
     }
 
+    /// Resolve reference flags for the current ast node.
+    fn resolve_reference_usages(&self) -> ReferenceFlag {
+        let mut flags = ReferenceFlag::None;
+
+        if self.nodes.parent_id(self.current_node_id).is_none() {
+            return ReferenceFlag::Read;
+        }
+
+        for parent in self.nodes.iter_parents(self.current_node_id) {
+            match parent.kind() {
+                AstKind::AssignmentTarget(_) | AstKind::SimpleAssignmentTarget(_) => {
+                    flags |= ReferenceFlag::Write;
+                    // continue up tree
```

Comment:
Is there a case where we need to go up from here?

Reply:
> Is there a case where we need to go up from here?

Did you find a case where we still need to go up from assignments? Theoretically we don't need to because JavaScript grammar doesn't have anything above assignment expressions (from the spec).

Reply:
Yes. UpdateExpressions come above AssignmentTargets